File: ops_observer/api_service/app/models/ops.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


HealthStatus = Literal["healthy", "degraded", "offline", "unknown"]
AlertSeverity = Literal["info", "warning", "critical"]
AlertStatus = Literal["open", "closed"]
ModuleType = Literal["gateway", "backend"]


class ModuleHealthSummary(BaseModel):
    module_id: str
    module_type: ModuleType
    display_name: str | None = None
    online: bool
    health_status: HealthStatus
    checked_at: str
    component_total: int = Field(default=0, ge=0)
    healthy_components: int = Field(default=0, ge=0)
    degraded_components: int = Field(default=0, ge=0)
    offline_components: int = Field(default=0, ge=0)
    last_error: str | None = None


class ModuleComponent(BaseModel):
    module_id: str
    component_id: str
    component_type: str
    display_name: str
    online: bool
    health_status: HealthStatus
    checked_at: str
    endpoint: str | None = None
    latency_ms: int | None = Field(default=None, ge=0)
    detail: str | None = None


class ModuleStats(BaseModel):
    module_id: str
    module_type: ModuleType
    updated_at: str
    data: dict[str, Any]


class OpsAlertRecord(BaseModel):
    id: int
    module_id: str
    module_type: ModuleType
    alert_type: str
    severity: AlertSeverity
    status: AlertStatus
    title: str
    detail: str
    created_at: str
    updated_at: str
    closed_at: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)


class OpsHealthListResponse(BaseModel):
    items: list[ModuleHealthSummary]
    updated_at: str | None = None


class OpsHealthDetailResponse(BaseModel):
    summary: ModuleHealthSummary
    components: list[ModuleComponent]
    stats: ModuleStats | None = None


class OpsAlertListResponse(BaseModel):
    items: list[OpsAlertRecord]


class OpsStatsResponse(BaseModel):
    items: list[ModuleStats]
    updated_at: str | None = None


class OpsSnapshotMessage(BaseModel):
    type: Literal["ops_snapshot"] = "ops_snapshot"
    data: OpsHealthListResponse


class OpsAlertMessage(BaseModel):
    type: Literal["ops_alert"] = "ops_alert"
    data: OpsAlertRecord

File: web/portal_app/src/types/ops.ts
export type HealthStatus = 'healthy' | 'degraded' | 'offline' | 'unknown';
export type AlertSeverity = 'info' | 'warning' | 'critical';
export type AlertStatus = 'open' | 'closed';

export interface ModuleHealthSummary {
  module_id: string;
  module_type: 'gateway' | 'backend';
  display_name?: string | null;
  online: boolean;
  health_status: HealthStatus;
  checked_at: string;
  component_total: number;
  healthy_components: number;
  degraded_components: number;
  offline_components: number;
  last_error?: string | null;
}

export interface ModuleComponent {
  module_id: string;
  component_id: string;
  component_type: string;
  display_name: string;
  online: boolean;
  health_status: HealthStatus;
  checked_at: string;
  endpoint?: string | null;
  latency_ms?: number | null;
  detail?: string | null;
}

export interface ModuleStats {
  module_id: string;
  module_type: 'gateway' | 'backend';
  updated_at: string;
  data: Record<string, unknown>;
}

export interface OpsAlertRecord {
  id: number;
  module_id: string;
  module_type: 'gateway' | 'backend';
  alert_type: string;
  severity: AlertSeverity;
  status: AlertStatus;
  title: string;
  detail: string;
  created_at: string;
  updated_at: string;
  closed_at?: string | null;
  payload: Record<string, unknown>;
}

export interface OpsHealthListResponse {
  items: ModuleHealthSummary[];
  updated_at?: string | null;
}

export interface OpsHealthDetailResponse {
  summary: ModuleHealthSummary;
  components: ModuleComponent[];
  stats?: ModuleStats | null;
}

export interface OpsAlertListResponse {
  items: OpsAlertRecord[];
}

export interface OpsStatsResponse {
  items: ModuleStats[];
  updated_at?: string | null;
}

export interface OpsMessage<T = unknown> {
  type: string;
  data: T;
}
